### PR TITLE
newrelic-infra-agent 1.20.0 (new formula)

### PR DESCRIPTION
**Pull request by brushknight**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

For now we use not final version of our code (pre-releases) in order to modify our code. Once we have everything ✅ on the pipeline, we will make a final release and change link & shasum for the tar.gz.

### CODE REVIEW

File: Formula/newrelic-infra-agent.rb
Comment (chenrui333) [inline Formula/newrelic-infra-agent.rb:9]: Why can't it use go 1.16?

File: Formula/newrelic-infra-agent.rb
Comment (SMillerDev) [inline Formula/newrelic-infra-agent.rb:38]:
```suggestion
  service do
    run [bin/"newrelic-infra-service", "-config", etc/"newrelic-infra/newrelic-infra.yml"]
    log_path var/"log/newrelic-infra/newrelic-infra.log"
    error_log_path var/"log/newrelic-infra/newrelic-infra.stderr.log"
  end
```

File: Formula/newrelic-infra-agent.rb
Comment (SMillerDev) [inline Formula/newrelic-infra-agent.rb:61]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:9]: This is our current tested version of go. In order to upgrade go we need to do a test for all features of the agent and it will take quite some time. We have this on the roadmap, but not a priority right now.

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:61]: This app is an infra-agent of NewRelic. Usually it starts, make a connection to the NR backend and start to send host telemetry data like CPU usage, memory usage, etc. Unfortunately we  don't have this easy to test flow yet. Can we continue without self test in this case and add it later?

https://github.com/newrelic/infrastructure-agent
https://docs.newrelic.com/docs/infrastructure/infrastructure-monitoring/get-started/get-started-infrastructure-monitoring/

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:38]: Thank you! I will update PR soon.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [reply Formula/newrelic-infra-agent.rb:9]: Per Homebrew policy, this should use the latest version of Go possible. We should only use earlier versions when the latest one is known to be broken.

**Comment by carlocab** [comment]
Please do `brew audit --strict --fix newrelic-infra-agent` to fix the CI errors.

Also, please don't mark steps in the PR template as done if you did not do them.

File: Formula/newrelic-infra-agent.rb
Comment (SMillerDev) [reply Formula/newrelic-infra-agent.rb:61]: > Can we continue without self test in this case and add it later?

No, the experience learns that if people don't write a good test at first they won't come back to fix it either.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [inline Formula/newrelic-infra-agent.rb:20]:
```suggestion
    goarch = Hardware::CPU.arm? ? "arm64" : "amd64"
    ENV["VERSION"] = version
    ENV["GOOS"] = "darwin"
    ENV["GOARCH"] = goarch
    ENV["CGO_ENABLED"] = "1"
    system "make", "dist-for-os"
    on_macos do
      bin.install "dist/darwin-newrelic-infra_darwin_#{goarch}/newrelic-infra"
      bin.install "dist/darwin-newrelic-infra-ctl_darwin_#{goarch}/newrelic-infra-ctl"
      bin.install "dist/darwin-newrelic-infra-service_darwin_#{goarch}/newrelic-infra-service"
```
We don't want to set `GOARCH=amd64` on ARM.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [inline Formula/newrelic-infra-agent.rb:13]: I don't think we want this on Linux.

Comment (SMillerDev) [reply Formula/newrelic-infra-agent.rb:13]: The install block below will also fail on linux

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:20]: Is it possible to bind this formula to darwin/amd64 only for now? We have m1 darwin on the roadmap and this is an MVP. For linux we have completely different flow, so we are not interested in linux version of brew now.

Comment (SMillerDev) [reply Formula/newrelic-infra-agent.rb:20]: > be maintained (i.e. the last release wasn’t ages ago, it works without patching on all Homebrew-supported OS versions and has no outstanding, unpatched security vulnerabilities)

https://docs.brew.sh/Acceptable-Formulae#niche-or-self-submitted-stuff

Comment (carlocab) [reply Formula/newrelic-infra-agent.rb:20]: We should get this to build on M1 and Linux. If we don't, this formula will make it on another Homebrew maintainer's to-do list.

I don't want to add a formula that I know will create work for another maintainer.

Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:20]: Just in case, do you know any way to test M1 mac (any cloud tested already) if you don't own physical one?

Comment (carlocab) [reply Formula/newrelic-infra-agent.rb:20]: AWS has M1 nodes, though they're expensive. Scaleway is a bit more reasonable.

Do you think this won't build on M1? Because if all you're compiling is Go code, then I don't think you have anything to worry about.

Comment (SMillerDev) [reply Formula/newrelic-infra-agent.rb:20]: Our sponsor macstadium has some: https://www.macstadium.com/m1-mini

Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:20]: > AWS has M1 nodes, though they're expensive. Scaleway is a bit more reasonable.

We checked AWS, there is only intel based bare metal options. Could you check which region did you find them? 
Another option (https://www.macstadium.com/m1-mini) will be complicated as we need to follow internal process of validation for each product we use and it can take months. 😞

> Do you think this won't build on M1? Because if all you're compiling is Go code, then I don't think you have anything to worry about.

For the m1 compilation, we need to validate that CGO dependent libraries will work as expected. 

For the linux, we will put some effort to bring this to the formula, likely we have amd64/arm64/arm fully tested there.

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb]: We are trying to make it work with linux, could you share documentation for this feature, as we can't find it in docs.

Comment (SMillerDev) [reply Formula/newrelic-infra-agent.rb]: `brew services` does not work with linux (yet). But it's here: https://github.com/Homebrew/homebrew-services/

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [reply Formula/newrelic-infra-agent.rb:20]: Ah, you're right. I was misinformed about AWS. My apologies.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [inline Formula/newrelic-infra-agent.rb:39]:
```suggestion
    output = shell_output("#{bin}/newrelic-infra -validate", 1)
    assert_match "config validation", output
```

Comment (brushknight) [reply Formula/newrelic-infra-agent.rb]: 👍

Comment (carlocab) [reply Formula/newrelic-infra-agent.rb]: Is `newrelic-infra -validate` not supposed to return an exit code of `1`?

Comment (brushknight) [reply Formula/newrelic-infra-agent.rb]: It could, we are adding this in the code right now, so we can change it a bit. Is it possible to have 1 or 0 accepted exit codes? If it validates and all good, should be 0, but if validates and finds errors, should be 1. 

For now we use 0 for both cases and print verbose message, but it is not ideal.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [inline Formula/newrelic-infra-agent.rb:14]:
```suggestion
    ENV["VERSION"] = version
```
Did this not work? If it gave you problems, we can do
```suggestion
    ENV["VERSION"] = version.to_s
```

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [inline Formula/newrelic-infra-agent.rb:9]:
```suggestion
  depends_on "go" => :build
  depends_on arch: :x86_64
```
It seems you're blocked on ARM by a dependency, so we can do this. Let's try to fix the Linux build, though.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [reply Formula/newrelic-infra-agent.rb]: I see; I thought I saw this return `1` on an earlier test run.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [inline Formula/newrelic-infra-agent.rb:27]: Move this into a `post_install` method. If you leave these in `install`, they'll just get cleaned up because they're empty.

File: Formula/newrelic-infra-agent.rb
Comment (SMillerDev) [inline Formula/newrelic-infra-agent.rb:38]: Does this exist on Linux?

File: Formula/newrelic-infra-agent.rb
Comment (SMillerDev) [inline Formula/newrelic-infra-agent.rb:31]: I think this is much easier.
```suggestion
    os = "darwin"
    ENV["CGO_ENABLED"] = "1"
    on_linux do
      os = "linux"
      ENV["CGO_ENABLED"] = "0"
    end

    ENV["GOOS"] = os
    system "make", "dist-for-os"
    bin.install "dist/#{os}-newrelic-infra_#{os}_#{goarch}/newrelic-infra"
      bin.install "dist/#{os}-newrelic-infra-ctl_#{os}_#{goarch}/newrelic-infra-ctl"
      bin.install "dist/#{os}-newrelic-infra-service_ #{os}_#{goarch}/newrelic-infra-service"
```

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:37]: Can we reach archive internals in post install part? To be able to do this:
```
(var/"db/newrelic-infra").install "assets/licence/LICENSE.macos.txt"
```

Comment (carlocab) [reply Formula/newrelic-infra-agent.rb:37]: No. That part you can keep in `install`, and there's no need to do `mkpath` with `var/"newrelic/db"`.

File: Formula/newrelic-infra-agent.rb
Comment (carlocab) [inline Formula/newrelic-infra-agent.rb:26]:
```suggestion
    on_macos do
```

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:14]: Will this version be fetched from archive url?

Comment (carlocab) [reply Formula/newrelic-infra-agent.rb:14]: Yes.

Comment (brushknight) [reply Formula/newrelic-infra-agent.rb]: fixed

**Review by carlocab — COMMENTED** [review]
Thanks, @brushknight. Just one small comment, and I think this can be merged when you mark 1.20 as your latest release on GitHub.

File: Formula/newrelic-infra-agent.rb
Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:10]: https://github.com/newrelic/infrastructure-agent/issues/695

Comment (carlocab) [reply Formula/newrelic-infra-agent.rb:11]: Thanks. We can merge this when 1.20 is your latest official release (this currently appears to be 1.19.4 on GitHub).

Comment (brushknight) [reply Formula/newrelic-infra-agent.rb:11]: Done. Could you verify it and merge, please?

**Review by carlocab — APPROVED** [review]
Thanks!

**Pull request closed without merging**
